### ISSUE TITLE
Research How to Host a Replica of Our Website

### ISSUE THREAD

**Issue by abenipa3** [hackforla/website#2014]
### Overview
We would like to invite designers to be more involved in reviewing pull requests but realized that it'll be a learning curve for some volunteers. 

Our team came up with a possible solution to build and run a replica of our website so that everyone can easily make changes to the replica without directly making changes to our actual website/our repo.

### Action Items
- [x] Research to see if it is feasible.
       - How do other companies do it?
       - How can we implement it in our team?
- [x] Research what are the steps?
- [x] Clone the Hack for LA Website and deploy to the app of choice. (As of 8/17/21, we chose Netlify.)
- [x] Test how the application works with forks.
- [ ] Discuss with the design and development team to see if this mock-up/prototype is workable.

### Resources/Instructions
- [Render](https://render.com/)
- [Travis-CI](https://www.travis-ci.com/)
- [Travis-CI Tutorial](https://docs.travis-ci.com/user/tutorial/)
- [Jekyll Doc on Travis CI](https://jekyllrb.com/docs/continuous-integration/travis-ci/)
- [squash.io](https://www.squash.io/)
- [Continuous Deployment with Netlify](https://docs.netlify.com/configure-builds/get-started/#build-settings)
- [Staging Environment with Netlify](https://www.unixtutorial.org/staging-environment-with-netlify/)
- [Migrating Your Jekyll Site to Netlify](https://www.netlify.com/blog/2017/05/11/migrating-your-jekyll-site-to-netlify/)
- [Documentation - About Netlify's Split Testing](https://docs.netlify.com/site-deploys/split-testing/)
- [YouTube - About Netlify's Split Testing](https://www.youtube.com/watch?v=r0ZA0zhLjkE)

**Comment by ExperimentsInHonesty** [comment]
@Aveline-art @alyssabenipayo please review closed issues looking for previous things we have tried and then reach out to CNK

**Comment by ExperimentsInHonesty** [comment]
@alyssabenipayo I tried to find any issues she was on related to this and could not.  Doesn't mean there was not one, but you could just reach out to her and she will tell you what we have tried. her github is cnk

**Comment by abenipa3** [comment]
> @alyssabenipayo I tried to find any issues she was on related to this and could not. Doesn't mean there was not one, but you could just reach out to her and she will tell you what we have tried. her github is cnk

Understood! I'll reach out to her then. Thanks, Bonnie!

**Comment by abenipa3** [comment]
@cnk  We're researching ways to build and run a replica of our website so that more volunteers would be more involved in pull request reviews, especially those who have little to no experience with Github.

So far I reviewed closed issues to see what we've tried and if there are other free and open-source automation servers apart from Travis-CI, Jenkins, and rendor.com, but not sure if I'm looking at the right issues to add to our resources.

May you please let us know what we have tried so far? Thank you!

**Comment by cnk** [comment]
We haven't tried anything towards this goal. 

Looking at Render, we might be able to create a GitHub Action that builds the site and pushes it to Render (or Netlify, or similar). I suspect the hard part for that would be the permissions to push the built files to a hosting service. 

Another project I work with uses Squash for just this sort of thing so I would probably start there: https://www.squash.io/#service I would hope we could keep to the limits of the free plan https://www.squash.io/pricing/

**Comment by abenipa3** [comment]
> We haven't tried anything towards this goal.
> 
> Looking at Render, we might be able to create a GitHub Action that builds the site and pushes it to Render (or Netlify, or similar). I suspect the hard part for that would be the permissions to push the built files to a hosting service.
> 
> Another project I work with uses Squash for just this sort of thing so I would probably start there: https://www.squash.io/#service I would hope we could keep to the limits of the free plan https://www.squash.io/pricing/

Thank you for your message, Cynthia! I'll research Squash to see how we could work towards the goal.

**Comment by abenipa3** [comment]
Progress: Reading about using the Squash deployment interface with Docker. Will start to document findings and upload them on HfLA's Drive.
Blockers: Focusing on interviews this week. Other than that, none.
Availability: Open from Tuesday night through Friday.
ETA: 8/6

**Comment by abenipa3** [comment]
Progress:

- Tested on squash.io and Travis CI to compare which is most feasible.
- Learned that netlify may be an alternative option for building a staging environment for our website. [[Sources](https://eduardoboucas.com/blog/2017/02/22/jekyll-staging-environment.html)]

Blockers:

- I realized this issue requires a bit of a learning curve on my end, including understanding how .travis.yml works. 
- Used up the free credits of squash.io when testing deployment.
- Question: Where is the best place in the hfla drive to share my findings? Should I create a new folder for this issue?

Availability: 8 hours

ETA: 8/18/21 (or sooner)

**Comment by Aveline-art** [comment]
@alyssabenipayo To answer your question: you can create a document in the drive with your findings, and I trust you will find an appropriate place to house this information. That said, I also prefer the findings to be documented here, if possible, and if not, at least linked at the top issue.

I have also looked into, briefly, at the GitHub API to see if it would be possible to write a program that clones, uploads, and delete forked repos. I can softly confirm that it might be possible. If none of the third-party apps work out, we can pursue that as our last option. If we reach that point, you have my blessing to return this issue into circulation with all of your typed notes unless you want to also tackle the challenge of researching and testing GH API.

**Comment by abenipa3** [comment]
@Aveline-art Thanks for your feedback! Here is a documentation of my findings below.

## Findings

<details> <summary> squash.io </summary>

- Steps I took to Test and Research:
   - Created an account.
   - Selected “New Deployment.”
- Testing:
   - Tested the hfla/website forked to my repo and this is the result.
![image](https://user-images.githubusercontent.com/38295612/129712832-83b572cf-c586-4b6e-a327-5ba2bf122c25.png)
   - Decided to test with a clone of the Guide Pages Prototype.
- Results:
   - squash.io provides a time limit in how long the deployment will run before it expires.
   - Test did not succeed as seen in the logs below.

![image](https://user-images.githubusercontent.com/38295612/129712957-d60b57d6-ed70-40d0-9cba-917f14e919e1.png)

![image](https://user-images.githubusercontent.com/38295612/129712988-5ba17873-8526-446c-9662-7560d8f495eb.png)

- FYI:
   - The free version of this website allows 30 hours/month of deployment.
       - Once the 30 hours expire, it will take a month to reset.

</details> 

<details> <summary> travis-ci </summary>

- Steps I took to Test and Research:
    - Created an account.
    - Added a `.travis.yml` file to my prototype's repository to tell Travis CI what to do.
- Testing:
  - Tested the hfla/website forked to my repo and a couple of prototypes. All resulted in this:
![image](https://user-images.githubusercontent.com/38295612/129713418-07959668-61e0-42cb-b556-0647c9728a68.png)
  - I added a .travis.yml file to my prototypes (excluding the hfla/website as I was cautious about doing so even though it's in my repo), and pushed to my github repo. 
  - Results:
     - No progress. 
     - Researched why it was the case and found these links from the Travis-CI community [here](https://travis-ci.community/t/no-builds-for-this-repository-and-no-requests-for-a-travis-ci-org-organization-repo/11443) and [here](https://travis-ci.community/t/no-builds-for-this-repository/151/3) as well as the [Jekyll’s Docs on Travis-CI](https://jekyllrb.com/docs/continuous-integration/travis-ci/). 
     - Still didn’t get the help needed, so I moved on to the next option (netlify).

</details> 

<details> <summary> netlify </summary>

- Must have a netlify account.
- Select “New Site from Git”
- Under Continuous Deployment, select the repository you want to deploy.
- Testing:
   - Tested with a clone of the Guide Pages Prototype.
   - Setting up was straightforward. Must have a Gemfile to have a working build. Was able to migrate jekyll site to netlify with this [link](https://www.netlify.com/blog/2017/05/11/migrating-your-jekyll-site-to-netlify/).
- Results:
   - Total deploy time: 20s.
   - Succeeded in deploying the cloned website [here](https://pages-clone-test.netlify.app/).
   - _One does not simply deploy a cloned website without encountering bugs, however..._
   - Stylings were missing as seen below. 
![image](https://user-images.githubusercontent.com/38295612/129714705-6663511d-ec7b-48ac-a516-340a5e00c820.png)
   - The links to the cloned guide pages lead to an error page. 
![image](https://user-images.githubusercontent.com/38295612/129714792-05eaf8f6-e2f3-4114-bfe3-1e6f8c1b7213.png)
   - To navigate to the working links of the individual guide pages, must delete  /pages-clone-test/ in https://pages-clone-test.netlify.app/pages-clone-test/github-issues. Example: https://pages-clone-test.netlify.app/github-issues
   - Is it feasible?
       - So far, no. Unless we find out how to deploy a clone without the errors mentioned above.
       - However, I would like to know our team's thoughts on this to see we could work through the errors.


</details> 

On another note, I feel reassured when you mentioned GitHub API because I have briefly read about it [here](https://docs.github.com/en/developers/overview/viewing-deployment-history) (as well as [Environments](https://docs.github.com/en/actions/reference/environments)) when I was looking for alternatives. I haven't done any testings after that, but it sounds like a workable solution. 🙌 

Please feel free to let me know if you have any questions. I'm one to discuss it during the dev meeting as well.

**Comment by cnk** [comment]
The lack of style on the Netlify site is because the style sheet and JS references also have the 'extra' part in the url. The stylesheet is at https://pages-clone-test.netlify.app/style.css but the reference to it from the web page is https://pages-clone-test.netlify.app/pages-clone-test/style.css.

Where is "pages-clone-test" coming from? is that the branch name?

**Comment by abenipa3** [comment]
> The lack of style on the Netlify site is because the style sheet and JS references also have the 'extra' part in the url. The stylesheet is at https://pages-clone-test.netlify.app/style.css but the reference to it from the web page is https://pages-clone-test.netlify.app/pages-clone-test/style.css.
> 
> Where is "pages-clone-test" coming from? is that the branch name?

Thanks for your feedback, Cynthia! I updated it by deleting the base URL from the prototype's config.yml and it works now.

https://pages-clone-test.netlify.app/

The "pages-clone-test" is a prototype I made from my repository for testing.

**Comment by abenipa3** [comment]
Update from this meeting 8/17/21:

Comments:
- We're not sure how to resolve the issue with Travis CI. We discussed adding a Gem File to it, which I've initially done, but still didn't work.
   - Note: If anyone has any input or experience in Travis CI, please let us know.
- Github API will require programming and may take some time to build.
- Squash.io will be our last option since 30 hours a month will not be feasible in case devs want to use it for longer.
- We will keep Travis CI, Github API, and Squash.io (last option).
- We decided to continue testing with Netlify. 

Action Item:

- [x] The next step is to `git clone` the HFLA website (from my repo).
- [x] Deploy it onto Netlify.
- [x]  @Aveline-art will test hosing on TravisCI

**Comment by abenipa3** [comment]
- Progress:
     - Successfully deployed cloned website onto Netlify.
        - The only change made in the repo for it to work is adding a GemFile.
     - Will research and test how it works with forked repos.
- Blockers: None.
- Availability: 8
- ETA: 8/26/21
- Link: https://cloned-hfla-site.netlify.app/

**Comment by Aveline-art** [comment]
Notes on Travis CI deployment:

- Travis CI does not actually host the repo.
  - TCI actually only test for builds against certain configurations and allows some level of testing in between builds.
    - This is useful for the dev-team but not the design team. 
  - In order to host, it needs to [push the successful build to a hosting website](https://docs.travis-ci.com/user/deployment/pages/)...such as gh-pages.
  - This means that for our purposes, Travis CI will go through multiple steps only to use a resource we already have.
  - TCI is a deadend. 

![results](https://user-images.githubusercontent.com/62368440/130305950-67b6dbad-f385-4b50-979a-32ebccbb536b.png)


@alyssabenipayo it looks like netlify is our winner. Let me know how the fork test is working out! Am very excited!

I also want to add that another possible, if somewhat crazy solution could be to use GHA to host a user's fork on the user's GHPages site.

Note: to create the successful build, I followed the steps in [Jekyll's documentation ](https://jekyllrb.com/docs/continuous-integration/travis-ci/)(thank you @alyssabenipayo for the link). In total, I created 3 files: cibuild, .travis.yml, and Gemfile, all of which are copied from the documentation.

**Comment by abenipa3** [comment]
## Progress Report
Progress:

- Researched fork test with netlify. (How to deploy from forked repo?). Hit a blocker.
- Discovered Netlify's feature called _split testing_, where we could run a branch based test.
~~- Tested split testing. (More details in the next section of this update.)~~
- Made a new account for fork testing. More details in the next section of this update.

Blockers:
None. I was able to test with forked repo in another account. (Not involving split testing.)
~~- I hit a blocker after reading [this post](https://answers.netlify.com/t/how-can-i-deploy-from-someone-elses-github-repo/23510/2).~~
~~- However, I have a little bit of hope for the split testing feature.~~

Availability: 6

ETA: Ready to review.

## Findings

### TL;DR

- Netlify works so far.
- Tested by forking the cloned-hfla-site from my account to a new account. 
- Netlify automatically deploys all branches so long as the settings for deployment are adjusted accordingly.

<details> <summary> More Details in Deployment Settings </summary>

1 - Select Site Setttings.
![image](https://user-images.githubusercontent.com/38295612/131214074-02c549ac-4fa3-48b6-82c7-609d5b4e7c0c.png)

2 - Click on Build & Deploy > Select Continuous Deployment > Scroll to Branches > If "Deploy all branches pushed to repository" is not selected, click on edit settings and select “All.”

![image](https://user-images.githubusercontent.com/38295612/131214055-f55f1ea4-1520-429d-b630-9d885462c30d.png)


</details>

- Netlify bot appears in pull requests and provides a link of the deployed site with the changes.
- The "original" cloned website would remain the same until changes are merged.
- I also added more resources to this issue that may help for other issues.
    - Documentation - About Netlify's Split Testing
    - YouTube - About Netlify's Split Testing

~~### About Split Testing~~

~~- Enables users to run a branch based test.~~
~~- Set up client-side analytics tracking. (Not relevant to this issue, but may be useful for other issues involving Google Analytics to track visitor behaviors.)~~

<details> <summary> Steps taken in Testing Forked Repo </summary>

- Created a new GH account for testing purposes.
- Forked [repo of the cloned-hfla-site](https://github.com/alyssabenipayo/cloned-hfla-site).
- Followed the contributing.md from our website as a new member would when setting up their environment for HFLA website.
- Edit docker-compose.yml with _test-cloned_hfla_site.
- Made changes to guide-pages.html as an example, `git push` all changes, and [created a pull request](https://github.com/alyssabenipayo/cloned-hfla-site/pull/3) on my account.
- Successfully deployed the [cloned website with the tested-changes](https://deploy-preview-3--cloned-hfla-site.netlify.app/guide-pages/survey-reporting-dashboard-guide).

![image](https://user-images.githubusercontent.com/38295612/131213678-d4ccfb7a-96a4-40af-8536-a8285a0e859c.png)

- Netlify bot automatically provides a link of the deployed website with the changes as seen below. (Select "Details")
    
![image](https://user-images.githubusercontent.com/38295612/131212482-c0e5c57f-0b01-4681-98ee-d989f069ef56.png)

- The ["original" cloned hfla site](https://cloned-hfla-site.netlify.app/guide-pages/survey-reporting-dashboard-guide) remains the same until pull request merges.

![image](https://user-images.githubusercontent.com/38295612/131213715-3723c4b9-c657-4bdf-b550-538d8abaa0a5.png)

</details> 

### Final Comments

- Figuring out how deleting branches/forks would work.
- Noticed that I was able to deploy more than one branches that I research build minutes; Would like to discuss with the team on their thoughts of Netlify being our application as well as the [pricings](https://www.netlify.com/pricing/) so far...

## Questions:
- Would this be considered as a DevOps project?

**Comment by Abel-Zambrano** [comment]
@Aveline-art @alyssabenipayo will discuss further in a meeting.

**Comment by abenipa3** [comment]
# Deploying HfLA Live Website to Netlify

## Prerequisites
Note: Most of the links mentioned in this section contain visuals of the steps taken to fulfill this process.

- Create an account for Hack for LA by using GitHub or Email.
- Select ["New Site"](https://user-images.githubusercontent.com/38295612/132080562-7f51bcd0-f0d1-4485-a07b-86e80b1f2245.png) from Git. 
- The site will take you to the “Create a new site” page. Under [Continuous Deployment](https://user-images.githubusercontent.com/38295612/132077771-d0a05dd1-6f86-4bba-8bc3-754a84587a3a.png), select GitHub.


- Follow the instructions. Select the repository you want to link to your site on Netlify. 
   - If you can’t see your repo, [configure the Netlify app on GitHub](https://github.com/apps/netlify/installations/new).

- A message in a [yellow box](https://user-images.githubusercontent.com/38295612/132077797-f690335d-776c-4895-8554-27f3019107c0.png) will appear on the screen. 
    - In this case, we will need to add a **Gemfile** before we can proceed. 
    - It will contain two lines in the file as seen [here](https://user-images.githubusercontent.com/38295612/132078081-75a48466-6efa-4b0d-a563-fb366a08019c.png).

- When completed, select Deploy Site. Wait for the site to build. Duration takes approx. 7 seconds. 
    - A [message ](https://user-images.githubusercontent.com/38295612/132077858-682cc38e-df1e-412d-a0fa-51d8f667a98c.png)will indicate when the site has been successfully deployed.


## Limitations

- [Bandwidth Used](https://docs.netlify.com/accounts-and-billing/billing/#bandwidth-usage) 
   - Bandwidth counter increases when the number of users visit the site. [[Source]](https://answers.netlify.com/t/what-does-actually-count-in-bandwidth/28605/4)
      - If one user spams the website with visits, the counter would increase as Netlify serves content to that user.
   - Every CSS, JS file served by the website would count against the bandwidth quota.
- [Build Minutes Used](https://docs.netlify.com/accounts-and-billing/billing/#builds-usage)
    - Time spent uploading new or changed files resulting from the build does count in the build minutes. [[Source]](https://answers.netlify.com/t/what-does-consitute-build-time/3889/8)
    - The time spent deploying after building does not count. [[Source]](https://answers.netlify.com/t/what-does-consitute-build-time/3889/3)
    - To check builds status and manage build capacity, visit `https://app.netlify.com/teams/[netlify account name]/builds/`
- Concurrent Builds
    - If we plan to build more than one website, there is no additional charge beyond the plan price for any plan - Starter, Pro, or Business - to queue up builds beyond the concurrency limit. [[Source]](https://answers.netlify.com/t/how-do-concurrent-builds-work/20086/2)
- Members
  - Must [pay](https://user-images.githubusercontent.com/38295612/132078598-594d697f-32f4-4504-b0b7-65a466fca2e2.png) to add more than one member. 
  - In this case, we'll have one team member maintaining the netlify account to keep to the free plan.

## Our Goals with Netlify
- Encourage more team members to review pull requests using the cloned website.
- Use Netlify's free plan for our weekly demos (specifically our Sunday Team meetings) before merging changes.

## Additional Resources
- [How to configure the Netlify app on GitHub?](https://github.com/apps/netlify/installations/new) - How to add repo to Netlify.
- [Netlify Billing](https://docs.netlify.com/accounts-and-billing/billing/) - More details on Builds Usage, Bandwidth Usage, etc.
- [Pricing Plan](https://www.netlify.com/pricing/) - Added this as reference to what's included in the free plan.
- [Build Minutes Pricing FAQ](https://www.netlify.com/pricing/faq/) - More details on Build Minutes.

**Comment by Aveline-art** [comment]
@alyssabenipayo I have discussed with @ExperimentsInHonesty about the demo days and we have decided that it would end up being less work to have the website staged on a separate repo rather than pursue netlify. Going to close this issue. Thank you for all the work you have put into this! I have learned a lot from you and I hope you have learned a lot as well!

**Comment by ExperimentsInHonesty** [comment]
@Aveline-art This issue ended up in UAT.  Because that is where issues go when you close them.  So an extra manual step will be needed in future to move it to the done column

**Comment by Aveline-art** [comment]
@ExperimentsInHonesty On the main point, I can be mindful about sorting issues myself, but is there a guideline to sort research-type issues (or dev issues that does not involve PRs) into UAT or done? Will there ever be a vested interest in reviewing research-type issues? Or is UAT purely visual-based? But even if so, are some non-visual-based issues still of interest to review at UAT? I feel the team needs some clear guidelines to avoid either under or over sorting closed issues.

On a side note, @Aveline-art, this is a reminder to create an issue for a proof of concept for the development repo. Steps include, but not limited to:

- [ ] Working it out in my own repo
- [ ] Asking for the needed repos to be made for an hfla test
- [ ] Set up the repo's settings and env
